src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation. For a given function specified by the input
+ * global var, it generates a new function which computes the adjoints of the given function's
+ * inputs in respect of the function's output. Use require_grads to specify inputs which need to be
+ * differentiated instead of returning adjoints of all inputs.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(std::move(mod)), require_grads_(std::move(require_grads)) {}
+
+  Function FuncTransform(Function func) {
+    ICHECK(func->body->IsInstance<SeqExprNode>());
+
+    Array<Var> new_params;
+    for (Var param : func->params) {
+      Var new_param = Var(param->vid, param->shape(), param->checked_type(), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+    Expr new_body = VisitWithNewScope(func->body);
+
+    const SeqExprNode* seq_expr = new_body.as<SeqExprNode>();
+    // only support a single dataflow block
+    ICHECK(seq_expr->blocks.size() == 1);
+    ICHECK(seq_expr->blocks[0]->IsInstance<DataflowBlockNode>());
+    const DataflowBlockNode* block = seq_expr->blocks[0].as<DataflowBlockNode>();
+
+    builder_->BeginDataflowBlock();
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+
+    // create adjoint var for inputs
+    for (size_t i = 0; i < new_params.size(); ++i) {
+      if (require_grads_.empty() || std::find(require_grads_.begin(), require_grads_.end(),
+                                              func->params[i]) != require_grads_.end()) {
+        CreateAdjointVar(new_params[i], false);
+      } else {
+        CreateAdjointVar(new_params[i], true);
```

Comment:
One good habit might be annotating the constant values.
```suggestion
        CreateAdjointVar(new_params[i], /*is_datalfow_var=*/false);
      } else {
        CreateAdjointVar(new_params[i], /*is_datalfow_var=*/true);
```